lib/src/dependency.dart

```diff
@@ -4,119 +4,138 @@
-    if (data.entries.isEmpty) {
-      // TODO: provide list of supported keys?
+  if (data is Map) {
+    try {
+      return _fromMap(data);
+    } on ArgumentError catch (e) {
       throw new CheckedFromJsonException(
-          data, null, 'Dependency', 'Must provide at least one key.');
+          data, e.name, 'Dependency', e.message.toString());
     }
+  }
 
-    if (data.containsKey('sdk')) {
-      return new SdkDependency.fromData(data);
-    }
+  return null;
+}
 
-    if (data.entries.length > 1) {
-      throw new CheckedFromJsonException(
-          data,
-          data.keys.skip(1).first as String,
-          'Dependency',
-          'Expected only one key.');
-    }
+Dependency _fromMap(Map data) {
+  if (data.entries.isEmpty) {
+// TODO: provide list of supported keys?
+    throw new CheckedFromJsonException(
+        data, null, 'Dependency', 'Must provide at least one key.');
+  }
 
-    var entry = data.entries.single;
-    var key = entry.key as String;
+  if (data.entries.length > 1) {
+    throw new CheckedFromJsonException(data, data.keys.skip(1).first as String,
+        'Dependency', 'Expected only one key.');
+  }
 
-    if (entry.value == null) {
-      throw new CheckedFromJsonException(
-          data, key, 'Dependency', 'Cannot be null.');
-    }
+  var entry = data.entries.single;
+  var key = entry.key as String;
 
-    switch (key) {
-      case 'path':
-        return new PathDependency.fromData(entry.value);
-      case 'git':
-        return new GitDependency.fromData(entry.value);
-    }
+  if (entry.value == null) {
+    throw new CheckedFromJsonException(
+        data, key, 'Dependency', 'Cannot be null.');
+  }
 
-    return null;
+  switch (key) {
+    case 'path':
+      return new PathDependency.fromData(entry.value);
+    case 'git':
+      return new GitDependency.fromData(entry.value);
   }
 
+  return null;
+}
+
+abstract class Dependency {
+  Dependency._();
+
   String get _info;
 
   @override
   String toString() => '$runtimeType: $_info';
 }
 
+@JsonSerializable(createToJson: false)
 class SdkDependency extends Dependency {
-  final String name;
+  final String sdk;
+  @JsonKey(fromJson: _constraintFromString)
   final VersionConstraint version;
 
-  SdkDependency(this.name, {this.version}) : super._();
+  SdkDependency(this.sdk, {this.version}) : super._();
 
-  factory SdkDependency.fromData(Map data) {
-    VersionConstraint version;
-    if (data.containsKey('version')) {
-      version = new VersionConstraint.parse(data['version'] as String);
+  static SdkDependency tryFromData(Object data) {
+    if (data is Map && data.containsKey('sdk')) {
+      return _$SdkDependencyFromJson(data);
     }
-    return new SdkDependency(data['sdk'] as String, version: version);
+    return null;
   }
 
   @override
-  String get _info => name;
+  String get _info => sdk;
 }
 
+@JsonSerializable(createToJson: false)
 class GitDependency extends Dependency {
+  @JsonKey(fromJson: _parseUri)
```

Comment:
can you not do `Uri.parse` directly? (is it not considered a constant?)

Reply:
json_serializable doesn't support `Class.staticMethod` – yet!